### PR TITLE
[IOTDB-3557] Maintain online data nodes by load manager

### PR DESCRIPTION
**Pull request by RYH61**
see details : https://issues.apache.org/jira/browse/IOTDB-3557

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java [hunk 1]
```diff
@@ -110,28 +110,28 @@ public void addMetrics() {
...
               Map::size,
               Tag.NAME.toString(),
               "online");
     }
   }
 
-  /** @return true if the specific DataNode is now online */
-  public boolean isOnlineDataNode(TDataNodeLocation info) {
+  /** @return true if the specific DataNode is registered */
+  public boolean isRegisteredDataNode(TDataNodeLocation info) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:129]: `public boolean isRegisteredDataNode(TDataNodeLocation dataNodeLocation)`

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 2]
```diff
@@ -263,11 +263,11 @@ private void updateNodeLoadStatistic() {
...
 
   /**
-   * Send heartbeat requests to all the online DataNodes
+   * Send heartbeat requests to all the Registered DataNodes
    *
-   * @param onlineDataNodes DataNodes that currently online
+   * @param onlineDataNodes DataNodes that registered in cluster
    */
-  private void pingOnlineDataNodes(List<TDataNodeInfo> onlineDataNodes) {
+  private void pingRegisteredDataNodes(List<TDataNodeInfo> onlineDataNodes) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:270]: `private void pingRegisteredDataNodes(List<TDataNodeInfo> registeredDataNodes)`

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 3]
```diff
@@ -171,16 +171,16 @@ public int getTotalCpuCoreCount() {
...
-    return nodeInfo.getOnlineDataNodes(dataNodeId);
+  public List<TDataNodeInfo> getRegisteredDataNodes(int dataNodeId) {
+    return nodeInfo.getRegisteredDataNodes(dataNodeId);
   }
 
-  public List<TDataNodesInfo> getOnlineDataNodesInfoList() {
+  public List<TDataNodesInfo> getRegisteredDataNodesInfoList() {
     List<TDataNodesInfo> dataNodesLocations = new ArrayList<>();
-    List<TDataNodeInfo> onlineDataNodes = this.getOnlineDataNodes(-1);
+    List<TDataNodeInfo> onlineDataNodes = this.getRegisteredDataNodes(-1);
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:183]: `List<TDataNodeInfo> registeredDataNodes`

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 4]
```diff
@@ -387,7 +387,7 @@ public interface ChangeServerListener {
 
   public List<TSStatus> flush(TFlushReq req) {
     List<TDataNodeInfo> onlineDataNodes =
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:389]: `List<TDataNodeInfo> registeredDataNodes`

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java (hunk 1 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:129]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java (hunk 4 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:389]: fixed~

**Review by CRZbulabula — APPROVED** [review]
LGTM~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 5]
```diff
@@ -327,6 +327,17 @@ public List<TConfigNodeLocation> getOnlineConfigNodes() {
         .collect(Collectors.toList());
   }
 
+  public List<TDataNodeInfo> getOnlineDataNodes(int dataNodeId) {
+    return getNodeManager().getRegisteredDataNodes(dataNodeId).stream()
```
Comment (qiaojialin) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:331]: Do we need to filter from registeredDataNodes? or could we use the heartbeatCacheMap directly?

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/RegionBalancer.java [hunk 6]
```diff
@@ -78,15 +78,15 @@ public CreateRegionGroupsPlan genRegionsAllocationPlan(
...
+      if (registeredDataNodes.size() < replicationFactor) {
         throw new NotEnoughDataNodeException();
       }
 
       for (int i = 0; i < allotment; i++) {
         // Generate allocation plan
         TRegionReplicaSet newRegion =
             regionAllocator.allocateRegion(
-                onlineDataNodes,
+                registeredDataNodes,
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/RegionBalancer.java:89]: +1

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java (hunk 5 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:331]: heartbeatCacheMap only stores the datanodeId and the running status of the datanode，The information of TDataNodeInfo still needs to be obtained from the registeredDataNode

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/RegionBalancer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/balancer/RegionBalancer.java]: fixed~
`getOnlineDataNodes()` is used in testing `testGetAndCreateDataPartition()` and `testGetAndCreateSchemaPartition()`, but `heartbeatCacheMap` is empty, so add 
`if (heartbeatCacheMap.isEmpty()) {
   dataNodeInfos = getNodeManager().getRegisteredDataNodes(dataNodeId);
}`
in getOnlineDataNodes(int dataNodeId)

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**